### DISCUSSION TITLE
2025 Second Update

### DISCUSSION THREAD

**Discussion by mdekstrand** [lenskit/lkpy#871] [Announcements]
Second major feature update + bugfix release for the 2025 series! Many changes, see the milestone and [release notes](https://lkpy.lenskit.org/stable/releases/2025.html) for more.

## Merged PRs

* Simplify test workfow by @mdekstrand in https://github.com/lenskit/lkpy/pull/666
* Add StochasticTopNSampler to be correct version of "softmax" sampler by @mdekstrand in https://github.com/lenskit/lkpy/pull/667
* Add flexible Torch matrix factorizer by @mdekstrand in https://github.com/lenskit/lkpy/pull/668
* Export configuration classes from lenskit.basic by @mdekstrand in https://github.com/lenskit/lkpy/pull/672
* Improve logging in Ray cluster setups by @mdekstrand in https://github.com/lenskit/lkpy/pull/673
* Limit test users to only have users who had item in the train data. by @sushobhan2024 in https://github.com/lenskit/lkpy/pull/676
* Add train and recommend commands to the CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/677
* Improve NumPy -> Torch conversions for read-only data by @mdekstrand in https://github.com/lenskit/lkpy/pull/680
* Reduce excess matrix manipulations in item-KNN by @mdekstrand in https://github.com/lenskit/lkpy/pull/678
* Improve progress bar update API by @mdekstrand in https://github.com/lenskit/lkpy/pull/681
* Update FlexMF parameters to better defaults by @mdekstrand in https://github.com/lenskit/lkpy/pull/682
* Run Doctor under coverage by @mdekstrand in https://github.com/lenskit/lkpy/pull/683
* Use uv for development environments instead of pixi by @mdekstrand in https://github.com/lenskit/lkpy/pull/700
* Add lenskit.state module and move the ParameterContainer interface by @mdekstrand in https://github.com/lenskit/lkpy/pull/702
* Create and use separate ModelTrainer objects by @mdekstrand in https://github.com/lenskit/lkpy/pull/701
* Support training checkpointing in ALS by @mdekstrand in https://github.com/lenskit/lkpy/pull/704
* Add CPU & GPU groups by @mdekstrand in https://github.com/lenskit/lkpy/pull/706
* Improve logging and reduce sorting in dataset relationships by @mdekstrand in https://github.com/lenskit/lkpy/pull/707
* Improve debug logging and add xopen to CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/708
* Introduce WARP loss to FlexMF by @mdekstrand in https://github.com/lenskit/lkpy/pull/709
* Build with setuptools instead of hatch by @mdekstrand in https://github.com/lenskit/lkpy/pull/710
* Add better compression support to CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/712
* Update pre-commit hook versions by @mdekstrand in https://github.com/lenskit/lkpy/pull/713
* Add Rust extension infrastructure and accelerate ItemKNN by @mdekstrand in https://github.com/lenskit/lkpy/pull/715
* Replace just with invoke for development tasks by @mdekstrand in https://github.com/lenskit/lkpy/pull/716
* Add versions to dataset schemas by @mdekstrand in https://github.com/lenskit/lkpy/pull/721
* Add SparseRow extension type for CSRs in Arrow by @mdekstrand in https://github.com/lenskit/lkpy/pull/722
* Use Rust to accelerate negative sampling by @mdekstrand in https://github.com/lenskit/lkpy/pull/723
* Re-add progress bars and signal checking to Item KNN builder by @mdekstrand in https://github.com/lenskit/lkpy/pull/724
* Add ndarray dep and use it for sampling by @mdekstrand in https://github.com/lenskit/lkpy/pull/725
* Use Maturin to simplify build setup by @mdekstrand in https://github.com/lenskit/lkpy/pull/726
* Reduce base structure memory consumption by @mdekstrand in https://github.com/lenskit/lkpy/pull/728
* Add Rust acceleration to User KNN by @mdekstrand in https://github.com/lenskit/lkpy/pull/730
* Rewrite ALS solvers in Rust by @mdekstrand in https://github.com/lenskit/lkpy/pull/731
* Rewrite FunkSVD in Rust by @mdekstrand in https://github.com/lenskit/lkpy/pull/732
* Build with ABI3 and maturin-action by @mdekstrand in https://github.com/lenskit/lkpy/pull/733
* Add SPEC-1 (Lazy Loading) and incorporate version numbers into docs by @mdekstrand in https://github.com/lenskit/lkpy/pull/736
* Implement is_sorted for matrix tables in Rust by @mdekstrand in https://github.com/lenskit/lkpy/pull/737
* Correctly detect no-genre movies in MovieLens by @mdekstrand in https://github.com/lenskit/lkpy/pull/738
* Adding Mean Average Precision by @albus-droid in https://github.com/lenskit/lkpy/pull/739
* Add tests to avoid overflow in softmax stochastic sampling by @mdekstrand in https://github.com/lenskit/lkpy/pull/742
* Add configurable weighting models by @mdekstrand in https://github.com/lenskit/lkpy/pull/744
* Parallelize item-item scoring by @mdekstrand in https://github.com/lenskit/lkpy/pull/745
* Support comma-separated lists for uniform nested parallelism configuration by @mdekstrand in https://github.com/lenskit/lkpy/pull/746
* Test and fix crash on Linux ARM by @mdekstrand in https://github.com/lenskit/lkpy/pull/748
* Update ALS solve to zero rows for no-info users/items by @mdekstrand in https://github.com/lenskit/lkpy/pull/749
* Add importers for UCSD Amazon data by @mdekstrand in https://github.com/lenskit/lkpy/pull/751
* refactor FlexMF with less string compares by @mdekstrand in https://github.com/lenskit/lkpy/pull/752
* Add sklearn NMF as a scorer by @FroggoLight in https://github.com/lenskit/lkpy/pull/741
* Speed up ItemList isin checks and evaluation metrics by @mdekstrand in https://github.com/lenskit/lkpy/pull/753
* Speed up ItemList indexing by @mdekstrand in https://github.com/lenskit/lkpy/pull/754
* Add Gini coefficient metrics by @mdekstrand in https://github.com/lenskit/lkpy/pull/755
* Speed up weight computations by @mdekstrand in https://github.com/lenskit/lkpy/pull/756
* Speed up weighted metrics by @mdekstrand in https://github.com/lenskit/lkpy/pull/757
* Add ItemList.top_n and speed up ItemList operations by @mdekstrand in https://github.com/lenskit/lkpy/pull/758
* Add subprocess progress reporting by @mdekstrand in https://github.com/lenskit/lkpy/pull/759
* Add a rate limiter and use ray.wait to manage Ray concurrency by @mdekstrand in https://github.com/lenskit/lkpy/pull/763
* Switch back to Just + docopt by @mdekstrand in https://github.com/lenskit/lkpy/pull/764
* Add lenskit data subset command, improve ItemListCollection conversion, and support random seeds in the CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/773
* Force Python UTF-8 mode for license hader update in pre-commit by @mdekstrand in https://github.com/lenskit/lkpy/pull/774
* Refactored batch accuracy tests into standard suite by @FroggoLight in https://github.com/lenskit/lkpy/pull/776
* Use pytest-retry to retry the flaky stochastic ranker test by @mdekstrand in https://github.com/lenskit/lkpy/pull/783
* Add component input hooks to the pipeline by @mdekstrand in https://github.com/lenskit/lkpy/pull/782
* Add additional tests to verify training item filtering by @mdekstrand in https://github.com/lenskit/lkpy/pull/788
* Use 64-bit row pointers in matrix relationship sets by @mdekstrand in https://github.com/lenskit/lkpy/pull/789
* Improve dataset documentation by @mdekstrand in https://github.com/lenskit/lkpy/pull/790
* Increaes minimum PyArrow vesion by @mdekstrand in https://github.com/lenskit/lkpy/pull/791
* Add LensKit settings and power querying by @mdekstrand in https://github.com/lenskit/lkpy/pull/794
* Revise documentation to be main site by @mdekstrand in https://github.com/lenskit/lkpy/pull/798
* Add devcontainer and update development dependencies by @mdekstrand in https://github.com/lenskit/lkpy/pull/799
* Automatically search up for lenskit.toml in CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/802
* Fix syntax in testsetup for item-list docs by @mdekstrand in https://github.com/lenskit/lkpy/pull/805
* added binarize_ratings function in DatasetBuilder by @vibhor-5 in https://github.com/lenskit/lkpy/pull/803
* Fix locate_configuration_root and Conda tests by @mdekstrand in https://github.com/lenskit/lkpy/pull/808
* Drop PyTorch max and update locked dependencies by @mdekstrand in https://github.com/lenskit/lkpy/pull/807
* Work around Ray extension type serialization errors by @mdekstrand in https://github.com/lenskit/lkpy/pull/809
* Use Rust-based hash table instead of Pandas index to implement Vocabulary by @mdekstrand in https://github.com/lenskit/lkpy/pull/811
* Refactor indirect hashtables to support reuse by @mdekstrand in https://github.com/lenskit/lkpy/pull/813
* Use uvx --isolated for compat tests by @mdekstrand in https://github.com/lenskit/lkpy/pull/816
* Use mise tasks instead of just by @mdekstrand in https://github.com/lenskit/lkpy/pull/814
* Implement a CoordinateTable structure for fast duplicate checking by @mdekstrand in https://github.com/lenskit/lkpy/pull/815
* Improve recommend CLI for batch inference, output, profiling by @mdekstrand in https://github.com/lenskit/lkpy/pull/817
* Fix coordinate table breakage when adding interactions after filtering or removing by @mdekstrand in https://github.com/lenskit/lkpy/pull/820
* Simplify PyPI CUDA setup by @mdekstrand in https://github.com/lenskit/lkpy/pull/824
* Implement usable pipeline configuration files and CLI support by @mdekstrand in https://github.com/lenskit/lkpy/pull/831
* Implement faster negative sampling and use CoordinateTable by @mdekstrand in https://github.com/lenskit/lkpy/pull/832
* More FlexMF performance improvements by @mdekstrand in https://github.com/lenskit/lkpy/pull/840
* Add parallel config to lenskit doctor by @mdekstrand in https://github.com/lenskit/lkpy/pull/842
* Add support for repeated interactions  by @FroggoLight in https://github.com/lenskit/lkpy/pull/812
* Make list metric DF have a standard Index when there is one key field by @mdekstrand in https://github.com/lenskit/lkpy/pull/845
* Add torch-geometric and LightGCN by @mdekstrand in https://github.com/lenskit/lkpy/pull/822
* Unify list and decomposed metric interfaces by @sushobhan2024 in https://github.com/lenskit/lkpy/pull/792
* Fix: correct attribute-to-entity mapping by @samiravaez in https://github.com/lenskit/lkpy/pull/850
* Fix Conda builds + GNN tests by @mdekstrand in https://github.com/lenskit/lkpy/pull/852
* Fix configuration warning on pipelines with placeholders by @mdekstrand in https://github.com/lenskit/lkpy/pull/853
* Update and fix pre-commit hooks by @mdekstrand in https://github.com/lenskit/lkpy/pull/855
* Refactor embedding size configuration for consistency + exponential support by @mdekstrand in https://github.com/lenskit/lkpy/pull/851
* Add validation for entity class name and attribute name  by @FroggoLight in https://github.com/lenskit/lkpy/pull/854
* Defer component instantiation when loading configurations by @mdekstrand in https://github.com/lenskit/lkpy/pull/858
* Add basic data-splitting CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/859
* Add experimental hyperparameter tuning library and CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/857
* Enable CI tests for the hyperparameter tuning by @mdekstrand in https://github.com/lenskit/lkpy/pull/861
* ALS: fast-path very large scoring by @mdekstrand in https://github.com/lenskit/lkpy/pull/862
* Add --save-pipeline option to tune CLI by @mdekstrand in https://github.com/lenskit/lkpy/pull/863
* Add Fair Re-Ranker (FA*IR) by @samiravaez in https://github.com/lenskit/lkpy/pull/860
* Add tuning settings to LensKit config + support a GPU resource multiplier by @mdekstrand in https://github.com/lenskit/lkpy/pull/864
* Make tuning submodules private by @mdekstrand in https://github.com/lenskit/lkpy/pull/868
* Add tune extra + update release notes by @mdekstrand in https://github.com/lenskit/lkpy/pull/870

## New Contributors
* @sushobhan2024 made their first contribution in https://github.com/lenskit/lkpy/pull/676
* @albus-droid made their first contribution in https://github.com/lenskit/lkpy/pull/739
* @FroggoLight made their first contribution in https://github.com/lenskit/lkpy/pull/741
* @vibhor-5 made their first contribution in https://github.com/lenskit/lkpy/pull/803

**Full Changelog**: https://github.com/lenskit/lkpy/compare/v2025.2.0...v2025.3.0

<hr /><em>This discussion was created from the release <a href='https://github.com/lenskit/lkpy/releases/tag/v2025.3.0'>2025 Second Update</a>.</em>
